edb/pgsql/deltafts.py

```diff
@@ -0,0 +1,471 @@
+    sql_kwarg_exprs: Dict[str, str],
+    options: qlcompiler.CompilerOptions,
+    schema: s_schema.Schema,
+    context: sd.CommandContext,
+) -> dbops.Command:
+    subject = index.get_subject(schema)
+    assert isinstance(subject, s_indexes.IndexableSubject)
+
+    effective, has_overridden = s_indexes.get_effective_fts_index(
+        subject, schema
+    )
+
+    if index != effective:
+        return dbops.CommandGroup()
+
+    # When creating an index on a child that already has an fts index
+    # inherited from the parent, we don't need to create the index, but just
+    # update the populating expressions.
+    if has_overridden:
+        return _refresh_fts_document(index, options, schema, context)
+    else:
+        return _create_fts_document(
+            index,
+            root_code,
+            index_expr,
+            predicate_src,
+            sql_kwarg_exprs,
+            schema,
+            context,
+        )
+
+
+def delete_fts_index(
+    index: s_indexes.Index,
+    drop_index: dbops.Command,
+    options: qlcompiler.CompilerOptions,
+    schema: s_schema.Schema,
+    context: sd.CommandContext,
+) -> dbops.Command:
+    subject = index.get_subject(schema)
+    assert isinstance(subject, s_indexes.IndexableSubject)
+
+    effective, _ = s_indexes.get_effective_fts_index(subject, schema)
+
+    if not effective:
+        return _delete_fts_document(index, drop_index, schema, context)
+    else:
+        # effective index remains: don't drop the fts document
+
+        effective_subject = effective.get_subject(schema)
+        is_eff_on_direct_parent = effective_subject in subject.get_bases(
+            schema
+        ).objects(schema)
+
+        if is_eff_on_direct_parent:
+            return _refresh_fts_document(index, options, schema, context)
+        else:
+            return dbops.CommandGroup()
+
+
+def _create_fts_document(
+    index: s_indexes.Index,
+    root_code: str,
+    index_expr: irast.Set,
+    predicate_src: Optional[str],
+    sql_kwarg_exprs: Dict[str, str],
+    schema: s_schema.Schema,
+    context: sd.CommandContext,
+) -> dbops.Command:
+    subject = index.get_subject(schema)
+    assert isinstance(subject, s_types.Type)
+
+    table_name = common.get_backend_name(schema, subject, catenate=False)
+
+    module_name = index.get_name(schema).module
+    index_name = common.get_index_backend_name(
+        index.id, module_name, catenate=False
+    )
+
+    subject_id = irast.PathId.from_type(schema, subject)
+    sql_res = compiler.compile_ir_to_sql_tree(
+        index_expr,
+        singleton_mode=True,
+        external_rvars={
+            (subject_id, 'source'): pgast.RelRangeVar(
+                alias=pgast.Alias(aliasname='NEW'),
+                relation=pgast.Relation(name='NEW'),
+            )
+        },
+    )
+    exprs = astutils.maybe_unpack_row(sql_res.ast)
+
+    ops = dbops.CommandGroup()
+    from edb.common import debug
+
+    if debug.flags.zombodb:
```

Comment:
I'd probably split zombo into another function unless it depends on a ludicrous number of shared variables?